### PR TITLE
SE: Run for local functions

### LINKED ISSUE #7469: SE: Run for local functions
Run SE on local functions and update/add UTs where needed.

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Fixes #7469

 Runs also for local functions in Top-level statements

### CODE REVIEW

**Comment by mary-georgiou-sonarsource** [comment]
@pavel-mikula-sonarsource  Please take a first look.
Things still to do:
- Fix test for S5773
- Check if there are FNs that the change did not affect.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs [hunk 1]
```diff
@@ -69,6 +69,14 @@ protected override void Initialize(SonarAnalysisContext context)
             SyntaxKind.OperatorDeclaration,
             SyntaxKind.MethodDeclaration);
 
+        context.RegisterNodeAction(
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs:72]: We need to test this in `SymblicExecutionRunnerTest.cs`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Conditions.cs [hunk 2]
```diff
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Net;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Conditions.cs:2]: This doesn't seem intentional. Missing selfreview?
```suggestion
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs [hunk 3]
```diff
@@ -69,6 +69,14 @@ protected override void Initialize(SonarAnalysisContext context)
...
+            c => Analyze<SyntaxNode>(context, c, x =>
+            {
+                var localFunction = (LocalFunctionStatementSyntaxWrapper)x;
+                return (SyntaxNode)localFunction.Body ?? localFunction.ExpressionBody;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs:76]: Both, `Body` and `ExpressionBody` will  need to be tested in the runner test above

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
This needs more specific testing

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Conditions.cs (hunk 2 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Conditions.cs:2]: Just missed it.

**Comment by mary-georgiou-sonarsource** [comment]
@pavel-mikula-sonarsource round 2.
I was thinking to add a test for functions in top-level statements with the top-level statements.

I'm also baffled a bit with that IT change.
I did some tests without the registration for local functions and this method there is still analyzed. Why it detects now there?

File: analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{
+"issues":  [
+{
+"id":  "S2259",
+"message":  "'e' is null on at least one execution path.",
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Remote\EndpointManager.cs",
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json:7]: If only C# would have been a human-readable language.

See line 550, it's a local function, inside a lambda, inside a method.
```
Directive Hopeless(HopelessAssociation e)
```

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json:7]: The readability problem here is that you need to look for tiny little syntax sugar clues that can be all over the place (beginning, middle, end of line) in unexpected locations. And they look all almost the same to each other without clear keywords that would make them easy to understand. And actually easy to find in global search tool too.

**Pull request merged**